File: nreduce/benchmark/mergesort2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char **genlist(int max)
{
  char **res = (char**)malloc(max*sizeof(char*));
  int i;
  for (i = 0; i < max; i++) {
    char tmp[100];
    char *rev;
    int len;
    int c;
    sprintf(tmp,"%d",7*i);
    len = strlen(tmp);
    rev = (char*)malloc(len+1);
    for (c = 0; c < len; c++)
      rev[c] = tmp[len-1-c];
    rev[len] = '\0';
    res[i] = rev;
  }
  return res;
}

void merge_sort(char **arr, char **aux, int lo, int hi)
{
  if (lo >= hi)
    return;

  int mid = lo+(hi+1-lo)/2;
  merge_sort(arr,aux,lo,mid-1);
  merge_sort(arr,aux,mid,hi);

  int pos = 0;
  int x = lo;
  int y = mid;
  while ((x <= mid-1) && (y <= hi)) {
    char *xval = arr[x];
    char *yval = arr[y];
    if (strcmp(xval,yval) < 0) {
      aux[pos++] = xval;
      x++;
    }
    else {
      aux[pos++] = yval;
      y++;
    }
  }
  if (x <= mid-1) {
    memcpy(&aux[pos],&arr[x],(mid-x)*sizeof(char*));
    pos += mid-x;
  }
  if (y <= hi) {
    memcpy(&aux[pos],&arr[y],(hi+1-y)*sizeof(char*));
    pos += hi+1-y;
  }
  memcpy(&arr[lo],aux,(hi+1-lo)*sizeof(char*));
}

int main(int argc, char **argv)
{
  int n = 1000;
  char **strings;
  char **temp;
  int i;
  if (2 <= argc)
    n = atoi(argv[1]);

  strings = genlist(n);
  temp = (char**)malloc(n*sizeof(char*));

  merge_sort(strings,temp,0,n-1);

  for (i = 0; i < n; i++)
    printf("%s\n",strings[i]);

  for (i = 0; i < n; i++)
    free(strings[i]);
  free(strings);
  free(temp);
  return 0;
}

File: nreduce/benchmark/mergesort2.pl
#!/usr/bin/perl -w

sub genlist
{
  my ($max) = @_;
  my @res = ();

  for (my $i = 0; $i < $max; $i++) {
    my $str = 7*$i;
    my $rev = reverse($str);
    push(@res,$rev);
  }

  return @res;
}

sub mergesort
{
  my ($arr,$aux,$lo,$hi) = @_;

  if ($lo >= $hi) {
    return;
  }

  my $mid = int($lo+($hi+1-$lo)/2);
  mergesort($arr,$aux,$lo,$mid-1);
  mergesort($arr,$aux,$mid,$hi);

  my $xlen = $mid-$lo;
  my $ylen = $hi+1-$mid;
  my $pos = 0;
  my $x = $lo;
  my $y = $mid;
  while (($x <= $mid-1) && ($y <= $hi)) {
    my $xval = $arr->[$x];
    my $yval = $arr->[$y];
    if ($xval lt $yval) {
      $aux->[$pos++] = $xval;
      $x++;
    }
    else {
      $aux->[$pos++] = $yval;
      $y++;
    }
  }

  my $i;

  if ($x <= $mid-1) {
    for ($i = 0; $i < $mid-$x; $i++) {
      $aux->[$pos+$i] = $arr->[$x+$i];
    }
    $pos += $mid-$x;
  }

  if ($y <= $hi) {
    for ($i = 0; $i < $hi+1-$y; $i++) {
      $aux->[$pos+$i] = $arr->[$y+$i];
    }
    $pos += $hi+1-$y;
  }
  for ($i = 0; $i < $hi+1-$lo; $i++) {
    $arr->[$lo+$i] = $aux->[$i];
  }
}


my $n = ($#ARGV < 0) ? 1000 : $ARGV[0];
my @strings = genlist($n);
my @temp = ();
mergesort(\@strings,\@temp,0,scalar(@strings)-1);
foreach $str (@strings) {
  print "$str\n";
}
